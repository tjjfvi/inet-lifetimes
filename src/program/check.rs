use crate::{
  display, err,
  error::{Error, ErrorGroup},
  globals::{ComponentInfo, GlobalCtx, PortLabel, TypeInfo},
  lifetimes::{LifetimeCtx, Side},
  order::{Order, Relation},
  program::{AgentDef, NetDef, Program, RuleDef, TypeDef},
};

impl Program {
  pub fn check(&mut self) -> ErrorGroup {
    let mut errors = ErrorGroup::default();

    for ty in &self.types {
      errors.push(ty.define(&mut self.globals));
    }

    for agent in &mut self.agents {
      errors.push(agent.define(&mut self.globals));
    }

    for net in &mut self.nets {
      errors.push(net.define(&mut self.globals));
    }

    for agent in self.agents.iter_mut() {
      if !self.globals.components.poisoned(agent.id) {
        errors.push(agent.check(&mut self.globals));
      }
    }

    errors
      .push(self.globals.type_order.check_coherent(|ty| self.globals.types.name(ty)).report("incoherent type order:"));

    for rule in self.rules.iter_mut() {
      errors.push(rule.check(&mut self.globals));
    }

    for net in self.nets.iter_mut() {
      if !self.globals.components.poisoned(net.id) {
        errors.push(net.check(&mut self.globals));
      }
    }

    errors
  }
}

impl TypeDef {
  fn define(&self, globals: &mut GlobalCtx) -> Result<(), Error> {
    globals.types.try_define(self.id, || TypeInfo { polarity: self.polarity })?;
    globals.types.try_define(!self.id, || TypeInfo { polarity: !self.polarity })?;
    Ok(())
  }
}

impl AgentDef {
  fn define(&mut self, globals: &mut GlobalCtx) -> Result<(), Error> {
    let mut errors = ErrorGroup::default();

    for port in &self.ports {
      errors.push(globals.types.get(port.0));
      errors.push(self.lt_ctx.lifetimes.get(port.1));
    }

    errors.push(self.lt_ctx.check_contract_satisfiable());

    errors.push(
      globals
        .components
        .try_define(self.id, || ComponentInfo { lt_ctx: self.lt_ctx.clone(), ports: self.ports.clone() }),
    );

    let name = globals.components.name(self.id);
    let ctx = display!("in agent `{name}`:");

    errors.report(ctx).map_err(|err| {
      globals.components.poison(self.id);
      err
    })
  }

  fn check(&mut self, globals: &mut GlobalCtx) -> Result<(), Error> {
    let mut required = Order::default();

    let pri = self.ports[0];
    for aux in &self.ports[1..] {
      if !aux.0 == pri.0 {
        required.relate_polarity(aux.1, pri.1, Relation::LT, globals.types[pri.0].polarity);
      } else {
        globals.type_order.relate(!aux.0, pri.0, Relation::LT);
      }
    }

    let mut full_order = self.lt_ctx.ex_order.clone();
    full_order.import(&self.lt_ctx.in_order, |x| x);

    let name = globals.components.name(self.id);
    let ctx = display!("in agent `{name}`:");

    self.lt_ctx.check_satisfiable(None, &full_order, &required).map_err(Error::context(ctx))
  }
}

impl RuleDef {
  fn check(&mut self, globals: &GlobalCtx) -> Result<(), Error> {
    let mut errors = ErrorGroup::default();

    let a_name = globals.components.name(self.a.component);
    let b_name = globals.components.name(self.a.component);
    let ctx = &display!("in rule `{a_name}-{b_name}`:");

    if self.a.ports[0] != self.b.ports[0] {
      errors.push(err!("matched nodes are not connected by their principal ports"));
    }

    let a = errors.push(globals.components.get(self.a.component));
    let b = errors.push(globals.components.get(self.b.component));

    for node in &self.result {
      errors.push(globals.components.get(node.component));
    }

    errors.report(ctx)?;

    let a = a.unwrap();
    let b = b.unwrap();

    let mut lt_ctx = LifetimeCtx::default();
    let a_base = lt_ctx.import(&a.lt_ctx, false, format_args!("{}.", a_name));
    let b_base = lt_ctx.import(&b.lt_ctx, false, format_args!("{}.", b_name));
    lt_ctx.ex_order.relate_polarity(
      a_base + a.ports[0].1,
      b_base + b.ports[0].1,
      Relation::LE,
      globals.types[a.ports[0].0].polarity,
    );

    for (lt_base, source_node) in [(a_base, &self.a), (b_base, &self.b)] {
      if let Some(pairs) = errors.push(self.var_ctx.check_node_arity(source_node, globals)) {
        for (i, (var, label)) in pairs.enumerate() {
          self.var_ctx.vars[var].uses.push(PortLabel(if i != 0 { !label.0 } else { label.0 }, lt_base + label.1))
        }
      }
    }

    self.var_ctx.infer_uses(&mut errors, &globals, &mut lt_ctx, &self.result);
    errors.push(self.var_ctx.check_types(&globals, &mut lt_ctx));

    errors.push(lt_ctx.check_satisfiable(Some(Side::Internal), &lt_ctx.ex_order, &lt_ctx.in_order));

    errors.report(ctx)
  }
}

impl NetDef {
  fn define(&mut self, globals: &mut GlobalCtx) -> Result<(), Error> {
    let mut errors = ErrorGroup::default();

    for (_, port) in &self.free_ports {
      errors.push(globals.types.get(port.0));
      errors.push(self.lt_ctx.lifetimes.get(port.1));
    }

    errors.push(self.lt_ctx.check_contract_satisfiable());

    errors.push(globals.components.try_define(self.id, || ComponentInfo {
      lt_ctx: self.lt_ctx.clone(),
      ports: self.free_ports.iter().map(|x| x.1).collect(),
    }));

    let name = globals.components.name(self.id);
    let ctx = display!("in net `{name}`:");

    errors.report(ctx).map_err(|err| {
      globals.components.poison(self.id);
      err
    })
  }

  fn check(&mut self, globals: &GlobalCtx) -> Result<(), Error> {
    let mut errors = ErrorGroup::default();

    let name = globals.components.name(self.id);
    let ctx = &display!("in net `{name}`:");

    for node in &self.nodes {
      errors.push(globals.components.get(node.component));
    }

    errors.report(ctx)?;

    for &(var, label) in &self.free_ports {
      self.var_ctx.vars[var].uses.push(PortLabel(!label.0, label.1))
    }

    self.var_ctx.infer_uses(&mut errors, &globals, &mut self.lt_ctx, &self.nodes);
    errors.push(self.var_ctx.check_types(&globals, &mut self.lt_ctx));

    errors.push(self.lt_ctx.check_satisfiable(Some(Side::Internal), &self.lt_ctx.ex_order, &self.lt_ctx.in_order));

    errors.report(ctx)
  }
}
