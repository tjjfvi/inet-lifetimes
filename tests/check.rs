use insta::{assert_snapshot, Settings};

use inet_lifetimes::check;

const OK_PATHS: &[&str] = &["examples/fn.inlt", "examples/list.inlt", "examples/nat.inlt"];

const ERR_PATHS: &[&str] = &["tests/programs/bad.inlt"];

#[test]
fn test_ok() {
  let mut settings = Settings::new();
  let mut err_count = 0;
  for path in OK_PATHS {
    settings.set_input_file(path);
    let result = check(path);
    if let Err(err) = result {
      println!("{path} failed: {err}");
      err_count += 1;
    }
  }
  if err_count != 0 {
    panic!("{err_count} failed")
  }
}

#[test]
fn test_err() {
  let mut settings = Settings::new();
  for &path in ERR_PATHS {
    settings.set_prepend_module_to_snapshot(false);
    settings.set_omit_expression(true);
    settings.set_input_file(path);
    let result = check(path).err().unwrap_or("no errors".to_owned());
    settings.bind(|| {
      assert_snapshot!(path, result);
    });
  }
}
