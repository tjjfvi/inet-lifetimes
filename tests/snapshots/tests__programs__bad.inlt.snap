---
source: tests/check.rs
input_file: tests/programs/bad.inlt
---
errors:
  duplicate definition of type `X`
  in agent `Foo`:
    previous error in type `X`
    undefined lifetime `'a`
  in agent `Foo`:
    undefined type `Y`
    undefined lifetime `'b`
    undefined lifetime `'b`
    duplicate definition of component `Foo`
  in net `x`:
    duplicate definition of component `x`
  in agent `Bar`:
    validity requires constraints not guaranteed:
      'b < 'a
  incoherent type order:
    A < C < B < A
  in net `foo`:
    `AB` expects 2 ports but 3 were supplied
    `BC` expects 2 ports but 1 were supplied
    `z`: mismatched types `C` and `!A`
    `c`: used only once
    `a`: used more than twice
